src/test/groovy/org/prebid/server/functional/pg/PgAuctionSpec.groovy

```diff
@@ -0,0 +1,576 @@
+
+        and: "Line items are fetched by PBS"
+        pgPbsService.sendForceDealsUpdateRequest(ForceDealsUpdateRequest.updateLineItemsRequest)
+
+        and: "User Service Response is set to return frequency capped id not identical to the line item fcapId"
+        userData.setUserDataResponse(UserDetailsResponse.defaultUserResponse.tap {
+            user.ext.fcapIds = [PBSUtils.randomNumber as String]
+        })
+
+        and: "Cookies header"
+        def uidsCookie = UidsCookie.defaultUidsCookie
+        def cookieHeader = HttpUtil.getCookieHeader(mapper, uidsCookie)
+
+        when: "Sending auction request to PBS"
+        def auctionResponse = pgPbsService.sendAuctionRequest(bidRequest, cookieHeader)
+
+        then: "PBS hasn't started processing PG deals as line item was recognized as frequency capped"
+        assert !auctionResponse.ext?.debug?.pgmetrics?.matchedTargetingFcapped
+        assert auctionResponse.ext?.debug?.pgmetrics?.readyToServe
+
+        cleanup:
+        userData.setUserDataResponse(UserDetailsResponse.defaultUserResponse)
+    }
+
+    def "PBS shouldn't use already matched line items by the same bidder during one auction"() {
+        given: "Bid request with two impressions"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            imp[0].ext.prebid.bidder = new Bidder(generic: new Generic())
+            imp << Imp.defaultImpression
+            imp[1].ext.prebid.bidder = new Bidder(generic: new Generic())
+        }
+        def accountId = bidRequest.site.publisher.id
+
+        and: "Bid response"
+        def bidResponse = BidResponse.getDefaultBidResponse(bidRequest)
+        bidder.setResponse(bidRequest.id, bidResponse)
+
+        and: "Planner Mock with two line items"
+        def plansResponse = PlansResponse.getDefaultPlansResponse(accountId).tap {
+            lineItems << LineItem.getDefaultLineItem(accountId)
+        }
+        generalPlanner.initPlansResponse(plansResponse)
+        def lineItemCount = plansResponse.lineItems.size()
+        def lineItemIds = plansResponse.lineItems.collect { it.lineItemId } as Set
+
+        and: "Line items are fetched by PBS"
+        pgPbsService.sendForceDealsUpdateRequest(ForceDealsUpdateRequest.updateLineItemsRequest)
+
+        when: "Auction is requested"
+        def auctionResponse = pgPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Two line items are ready to be served"
+        assert auctionResponse.ext?.debug?.pgmetrics?.readyToServe?.size() == lineItemCount
+
+        and: "Two (as the number of imps) different line items were sent to the bidder"
+        def sentToBidder = auctionResponse.ext?.debug?.pgmetrics?.sentToBidder?.get(GENERIC.value)
+        assert sentToBidder?.size() == lineItemCount
+        assert sentToBidder.sort() == lineItemIds.sort()
+
+        def sentToBidderAsTopMatch = auctionResponse.ext?.debug?.pgmetrics?.sentToBidderAsTopMatch?.get(GENERIC.value)
```

Comment:
Can we replace the key in sentToBidderAsTopMatch with BidderName?

Reply:
No, it may be used bidder alias in the request to PBS (it may be any string).